### ISSUE TITLE
Failed build when installing Nix package manager from source

### ISSUE THREAD

**Issue by tmward** [NixOS/nixpkgs#117389]
I am unable to install Nix package manager from source. I have installed all the listed [prerequisites](https://nixos.org/manual/nix/stable/#sec-prerequisites-source). I have tried with different compiler versions (I built gcc 10.2 from source since the provided gcc 9.2 on the cluster was having an error), but each one still produces the same error.

**To Reproduce**
1. `wget https://nixos.org/releases/nix/nix-2.3.10/nix-2.3.10.tar.xz`
2. `tar xf nix-2.3.10.tar.xz`
3. `cd nix-2.3.10`
4. `./configure --prefix=$HOME/nix`
5. `make`

which outputs (just the last tail of the output):

``` 
CXX    src/libutil/util.o
src/libutil/util.cc: In function 'void nix::_interrupted()':
src/libutil/util.cc:1182:35: warning: 'bool std::uncaught_exception()' is deprecated [-Wdeprecated-declarations]
 1182 |     if (!interruptThrown && !std::uncaught_exception()) {
      |                                   ^~~~~~~~~~~~~~~~~~
In file included from src/libutil/lazy.hh:1,
                 from src/libutil/util.cc:1:
/n/home03/tmw28/local/include/c++/10.2.0/exception:106:8: note: declared here
  106 |   bool uncaught_exception() _GLIBCXX_USE_NOEXCEPT __attribute__ ((__pure__));
      |        ^~~~~~~~~~~~~~~~~~
src/libutil/util.cc:1182:54: warning: 'bool std::uncaught_exception()' is deprecated [-Wdeprecated-declarations]
 1182 |     if (!interruptThrown && !std::uncaught_exception()) {
      |                                                      ^
In file included from src/libutil/lazy.hh:1,
                 from src/libutil/util.cc:1:
/n/home03/tmw28/local/include/c++/10.2.0/exception:106:8: note: declared here
  106 |   bool uncaught_exception() _GLIBCXX_USE_NOEXCEPT __attribute__ ((__pure__));
      |        ^~~~~~~~~~~~~~~~~~
  CXX    src/libutil/xml-writer.o
  LD     src/libutil/libnixutil.so
g++: error: unrecognized command-line option '-R'
g++: error: unrecognized command-line option '-R'
make: *** [mk/lib.mk:104: src/libutil/libnixutil.so] Error 1
```

Full `config.log` is attached:

[config.log](https://github.com/NixOS/nixpkgs/files/6191835/config.log)

I also recorded the entire build with `script(1)` which is attached as `nix_build_log.txt`:
[nix_build_log.txt](https://github.com/NixOS/nixpkgs/files/6191837/nix_build_log.txt)

**Expected behavior**
Successful build

**Additional context**
I am installing from source because I do not have root access on the cluster, so I need to change the installation prefix for Nix. The cluster software is older (CentOS 7). I tried to build Nix with the default (old versions) and a newer gcc (9.2) but this failed. So I went and installed every listed prerequisite from source to the most recent stable versions. Despite this, I still am receiving the same error ("unrecognized command-line option -R"). Please let me know if anymore information is needed. I greatly appreciate help with debugging this!

**Comment by veprbl** [comment]
I'm using https://git.sr.ht/~veprbl/nix-bootstrap to bootstrap nix on RHEL7-like systems.

**Comment by tmward** [comment]
Hi, thank you for the script! I've been plugging away with a lot of trial and error since you shared it and got things to almost work:

It worked unmodified until it tried to install release-20.03 which had an out-of-date SHA256 sum for one of the packages so it would not continue.

I then thought, ok, let's use the latest release, 20.09. This, however, required nix version >= 2.2. I modified your script to use the latest 2.3 series nix. However, this needed a gcc with c++17 support so the build failed (since gcc 5.5 did not have that support). So, I used nix 2.2.2. This needed sqlite and libeditline, so I added those to your build script. I also made it put  `sandbox = false` in `nix.conf` since the nix 2.2 and up default to using it (and namespaces aren't enabled on this cluster). I made a few other tweaks to your script (it tried to unpack gcc even if it was skipping gcc build so errored since it wasn't download, fixed anything that shellcheck didn't like, fixed nix-env which needed `--cores` not `-j` to speed up build time, sanitized more env variables to ensure it built without being contaminated, etc.) Here is the updated version:
[bootstrap.diff.txt](https://github.com/NixOS/nixpkgs/files/6206025/bootstrap.diff.txt)

After all that modification, and waiting for lots of building time, everything installed... except nix! It built, but the tests are failing (it's trying to use `unshare` and like I said above, user namespaces are not enabled). Here's the tail of the build output:

```
running test tests/function-trace.sh... [FAIL]
    common.sh: line 89: unshare: command not found
    + set +x
    Tracing expression 'builtins.tryEval (throw "example")' failed. difference:
    2,8d1
    < function-trace entered undefined position at
    < function-trace exited undefined position at
    < function-trace entered (string):1:1 at
    < function-trace entered (string):1:19 at
    < function-trace exited (string):1:19 at
    < function-trace exited (string):1:1 at
    < 
53 out of 56 tests failed 
make: *** [mk/tests.mk:12: installcheck] Error 1
builder for '/n/home03/tmw28/nix/store/5h9vjjmvlkiacsamm8qfvbdzzlr0s64m-nix-2.3.10.drv' failed with exit code 2
error: build of '/n/home03/tmw28/nix/store/5h9vjjmvlkiacsamm8qfvbdzzlr0s64m-nix-2.3.10.drv' failed
```

Do you know how to, when trying to install nix, pass along the "--disable-seccomp-sandboxing" flag to the build for nix? I'm building nix with the following command;

```
nix-env --cores 20 -iA nix -f ~/nixpkgs
```

and my `nix.conf` has `sandbox = false` in it already. However, the nix that I build during tests still is trying to use sandbox and therefore fails. I just need to pass "--disable-seccomp-sandboxing " to the configure for nix but can't figure out how to do that. Thank you for the help!

**Comment by veprbl** [comment]
@tmward 
Script is a bit outdated, indeed, but I'm still glad you could make use of it. Your changes look really great, I'd like to incorporate them into my distribution if you don't mind.

Regarding the linking with seccomp, I think you have two options: add an override `withLibseccomp = false;` to the nixpkgs/config.nix, or (what I do) is to have `filter-syscalls = false` in your nix.conf.

**Comment by tmward** [comment]
@veprbl of course, feel free to use the changes!

So I tried both your options with no success. I then did the nuclear option and hard-coded the configuration flag in `pkgs/tools/package-management/nix/default.nix`. Maddeningly, none of the above worked. It still kept on doing the install tests which required having user namespaces.

As a work-around, I added `doInstallCheck = false;` to my nixpkgs/config.nix to override attributes which allowed it to install.

Unfortunately, after installation, I get this running any of the nix commands (after exporting NIX_PATH and sourcing nix.sh):

```
nix: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

Anything immediately obvious where I messed up?
